README.md
# Dexter
VSCode extension for code completion using OpenAI's Codex 

Bring your own key


dexter/CHANGELOG.md
# Change Log

All notable changes to the "dexter" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

dexter/out/codex.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Codex = void 0;
const vscode = require("vscode");
class Codex {
    constructor(apiKey) {
        this.context = '';
        this.query = '';
        this.resLength = 64;
        this.temp = 0;
        this.topp = 1;
        this.freqPenalty = 0;
        this.presPenalty = 0;
        this.bestOf = 1;
        this.injectStart = '';
        this.injectRestart = '';
        this.engine = 'davinci-codex';
        this.key = apiKey;
    }
    addContext(context) {
        this.context = this.context + '\n' + context;
    }
    setEngine(engine) {
        this.engine = engine;
    }
    clearContext() {
        this.context = '';
    }
    getContext() {
        return this.context || "";
    }
    addQuery(query) {
        this.query = query;
    }
    clearQuery() {
        this.query = "";
    }
    setBestOf(bestOf) {
        this.bestOf = bestOf;
    }
    setResLength(length) {
        if (length >= 4096) {
            this.resLength = 4096;
        }
        else if (length <= 1) {
            this.resLength = 1;
        }
        else {
            this.resLength = length;
        }
    }
    setTemp(temp) {
        if (temp >= 1) {
            this.temp = 1;
        }
        else if (temp <= 0) {
            this.temp = 0;
        }
        else {
            this.temp = temp;
        }
    }
    setTopp(topp) {
        if (topp >= 1) {
            this.topp = 1;
        }
        else if (topp <= 0) {
            this.topp = 0;
        }
        else {
            this.topp = topp;
        }
    }
    setFreqPenalty(freqPenalty) {
        if (freqPenalty >= 1) {
            this.freqPenalty = 1;
        }
        else if (freqPenalty <= 0) {
            this.freqPenalty = 0;
        }
        else {
            this.freqPenalty = freqPenalty;
        }
    }
    setPresPenalty(presPenalty) {
        if (presPenalty >= 1) {
            this.presPenalty = 1;
        }
        else if (presPenalty <= 0) {
            this.presPenalty = 0;
        }
        else {
            this.presPenalty = presPenalty;
        }
    }
    setStopSequences(sequence) {
        this.stopSequences = sequence.split("/|").filter(function (val, index) {
            return index < 4;
        });
    }
    clearStopSequences() {
        this.stopSequences = [];
    }
    async complete(appendResultAndQueryToContext) {
        if (!this.context || !this.query) {
            vscode.window.showErrorMessage("Codex: No Query Provided");
            return '';
        }
        const headers = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.key}`
        };
        const data = {
            "prompt": this.context + '\n\n' + this.query,
            "max_tokens": this.resLength,
            "temperature": this.temp,
            "top_p": this.topp,
            "n": this.bestOf,
            "stream": false,
            "logprobs": null,
            "stop": this.stopSequences
        };
        await fetch('https://api.openai.com/v1/engines/davinci-codex/completions', {
            method: "POST",
            headers: headers,
            body: JSON.stringify(data)
        })
            .then((response) => {
            if (response.status === 200) {
                return response.json().then((body) => {
                    const result = body.choices[0].text;
                    if (appendResultAndQueryToContext) {
                        this.addContext(result);
                        this.clearQuery();
                    }
                    return result;
                }).catch();
            }
            ;
        }).catch(err => {
            vscode.window.showErrorMessage(err.message);
            return '';
        });
        return '';
    }
}
exports.Codex = Codex;
//# sourceMappingURL=codex.js.map

dexter/src/codex.ts
import * as vscode from 'vscode';

export class Codex {
    key: string;
    context?: string = '';
    queries?: string[] = [];
    resLength?: number = 64;
    temp?: number = 0;
    topp?: number = 1;
    freqPenalty?: number = 0;
    presPenalty?: number = 0;
    bestOf?: number = 1;
    stopSequences?: string[];
    injectStart?: string = '';
    injectRestart?: string = '';
    engine?: string = 'davinci-codex';
    
    constructor(apiKey: string) {
        this.key = apiKey;
    }

    addContext(context: string) {
        this.context = this.context + '\n' + context;
    }

    setEngine(engine: string) {
        this.engine = engine;
    }

    clearContext() {
        this.context = '';
    }

    getContext() {
        return this.context || "";
    }

    addQuery(query: string) {
        this.queries?.push(query) ;
    }

    clearQueries() {
        this.queries = [];
    }

    setBestOf(bestOf: number) {
        this.bestOf = bestOf;
    }

    setResLength(length: number) {
        if (length >= 4096) {
            this.resLength = 4096;
        } else if (length <= 1) {
            this.resLength = 1;
        } else {
            this.resLength = length;
        }
    }

    setTemp(temp: number) {
        if (temp >= 1) {
            this.temp = 1;
        } else if (temp <= 0) {
            this.temp = 0;
        } else {
            this.temp = temp;
        }
    }

    setTopp(topp: number) {
        if (topp >= 1) {
            this.topp = 1;
        } else if (topp <= 0) {
            this.topp = 0;
        } else {
            this.topp = topp;
        }
    }
    
    setFreqPenalty(freqPenalty: number) {
        if (freqPenalty >= 1) {
            this.freqPenalty = 1;
        } else if (freqPenalty <= 0) {
            this.freqPenalty = 0;
        } else {
            this.freqPenalty = freqPenalty;
        }
    }
    
    setPresPenalty(presPenalty: number) {
        if (presPenalty >= 1) {
            this.presPenalty = 1;
        } else if (presPenalty <= 0) {
            this.presPenalty = 0;
        } else {
            this.presPenalty = presPenalty;
        }
    }

    setStopSequences(sequence: string){
        if (sequence.length === 0) {
            this.stopSequences = [];
            return
        }
        this.stopSequences = sequence.split("/|").filter(function (val, index) {
            return index < 4;
        });
    }

    clearStopSequences() {
        this.stopSequences = [];
    }

    getStopSequences() {
        return this.stopSequences;
    }

    async complete() {
        const headers = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.key}`
        };

        var requests: Promise<Response>[] = [];

        this.queries?.forEach(async (query) => {
            const data = {
                "prompt": this.context + '\n\n' + query,
                "max_tokens": this.resLength,
                "temperature": this.temp,
                "top_p": this.topp,
                "n": this.bestOf,
                "stream": false,
                "logprobs": null,
                "stop": this.stopSequences
            }
            requests.push(fetch('https://api.openai.com/v1/engines/davinci-codex/completions', {
                method: "POST",
                headers: headers,
                body: JSON.stringify(data)
            }));
        });

        var strings: string[] = [];

        await Promise.all(requests).then((values) => {
            values.forEach(async (value) => {
                if (value.status === 200) {
                    await value.json().then((body) => {
                        strings.push(body.choices[0].text);
                    }).catch();
                } else if (value.status === 401) {
                    strings.push('UNAUTHORIZED');
                } else {
                    strings.push('ERROR');
                    await value.json().then((body) => {
                        strings.push(`${value.status} - ${body.error.message}`)
                    }).catch();
                }
            });           
        });

        this.clearQueries();
        return strings;
        
    }
}

dexter/out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
const vscode = require("vscode");
const codex_1 = require("./codex");
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
function activate(context) {
    let codex;
    // Use the console to output diagnostic information (console.log) and errors (console.error)
    // This line of code will only be executed once when your extension is activated
    console.log('Congratulations, your extension "dexter" is now active!');
    // The command has been defined in the package.json file
    // Now provide the implementation of the command with registerCommand
    // The commandId parameter must match the command field in package.json
    context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(cfg => {
        const config = vscode.workspace.getConfiguration('dexter');
        if (cfg.affectsConfiguration("dexter.apiKey")) {
            codex = new codex_1.Codex(config.get('apiKey', ''));
        }
        if (cfg.affectsConfiguration("dexter.engine")) {
            codex.setEngine(config.get("dexter.engine", ""));
        }
        if (cfg.affectsConfiguration("dexter.responseLength")) {
            codex.setResLength(config.get("dexter.responseLength", 64));
        }
        if (cfg.affectsConfiguration("dexter.temperature")) {
            codex.setTemp(config.get("dexter.temperature", 0));
        }
        if (cfg.affectsConfiguration("dexter.topP")) {
            codex.setTopp(config.get("dexter.topP", 1));
        }
        if (cfg.affectsConfiguration("dexter.frequencyPenalty")) {
            codex.setFreqPenalty(config.get("dexter.frequencyPenalty", 0));
        }
        if (cfg.affectsConfiguration("dexter.presencePenalty")) {
            codex.setPresPenalty(config.get("dexter.presencePenalty", 0));
        }
        if (cfg.affectsConfiguration("dexter.bestOf")) {
            codex.setBestOf(config.get("dexter.bestOf", 1));
        }
        if (cfg.affectsConfiguration("dexter.stopSequences")) {
            codex.setStopSequences(config.get("dexter.stopSequences", ""));
        }
    }));
    context.subscriptions.push(vscode.commands.registerCommand('dexter.addCodexContext', () => {
        const editor = vscode.window.activeTextEditor;
        if (!editor) {
            vscode.window.showInformationMessage("No open editor");
        }
        else {
            var selections = editor.selections;
            selections.forEach((selection, index) => {
                const text = editor.document.getText(selection);
                if (text.length > 0) {
                    codex.addContext(text);
                }
                else {
                    selections.splice(index, 1);
                    return;
                }
            });
            // vscode.window.showInformationMessage(`Added ${selections.length} ${selections.length === 1 ? 'line' : 'lines'}. Total: ${codex.getContext().split('\n')}`)
            vscode.window.showInformationMessage(`Added context from ${selections.length} selections`);
        }
    }));
    context.subscriptions.push(vscode.commands.registerCommand('dexter.clearCodex', () => {
        codex.clearContext();
        codex.clearQuery();
        codex.clearStopSequences();
    }));
    function appendText(text) {
        vscode.commands.registerTextEditorCommand('dexter.insertText', function (editor, edit, args) {
            edit.insert(editor.selection.end, text);
        });
    }
    function replaceText(text) {
        vscode.commands.registerTextEditorCommand('dexter.insertText', function (editor, edit, args) {
            edit.replace(editor.selection.end, text);
        });
    }
    context.subscriptions.push(vscode.commands.registerCommand('dexter.runCodeCompletion', () => {
        const key = vscode.workspace.getConfiguration('dexter').get('apiKey', false);
        if (!key) {
            vscode.window.showErrorMessage('API key must be specified in the extension settings');
        }
        else {
            const editor = vscode.window.activeTextEditor;
            if (!editor) {
                vscode.window.showInformationMessage("No open editor");
            }
            else {
                const selections = editor.selections;
                const replaceHighlight = vscode.workspace.getConfiguration('dexter').get('replaceHighlight');
                const sameCompletion = vscode.workspace.getConfiguration('dexter').get('sameCompletion');
                var insertText = replaceHighlight ? appendText : replaceText;
                //append if one or more selections contain characters
                var appendResult = !selections.every((value) => { return value.isEmpty; });
                codex.complete(appendResult).then(result => {
                    selections.forEach((selection, index) => {
                        codex.addQuery(editor.document.getText(selection));
                        if (index === 0 || sameCompletion) {
                            insertText(result);
                        }
                        else {
                            codex.complete(appendResult).then(result => {
                                insertText(result);
                            });
                        }
                    });
                    return;
                });
            }
        }
    }));
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map

dexter/src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { Codex } from './codex';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
	let codex: Codex;
	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	
	console.log('Congratulations, your extension "dexter" is now active!');

	// The command has been defined in the package.json file
	// Now provide the implementation of the command with registerCommand
	// The commandId parameter must match the command field in package.json

	context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(cfg => {
		//use else ifs if each change is recorded individually. if multiple changes occur at once, uses ifs
		const config = vscode.workspace.getConfiguration('dexter');
		if (cfg.affectsConfiguration("dexter.apiKey")) {
			codex = new Codex(config.get('apiKey', ''));
			codex.setStopSequences(config.get("dexter.stopSequences", ""));
			codex.setBestOf(config.get("dexter.bestOf", 1));
			codex.setPresPenalty(config.get("dexter.presencePenalty", 0));
			codex.setFreqPenalty(config.get("dexter.frequencyPenalty", 0));
			codex.setTopp(config.get("dexter.topP", 1));
			codex.setTemp(config.get("dexter.temperature", 0));
			codex.setResLength(config.get("dexter.responseLength", 64));
			codex.setEngine(config.get("dexter.engine", ""));
		}
		if (cfg.affectsConfiguration("dexter.engine")) {
			codex.setEngine(config.get("dexter.engine", ""));
		}
		if (cfg.affectsConfiguration("dexter.responseLength")) {
			codex.setResLength(config.get("dexter.responseLength", 64));
		}
		if (cfg.affectsConfiguration("dexter.temperature")) {
			codex.setTemp(config.get("dexter.temperature", 0));
		}
		if (cfg.affectsConfiguration("dexter.topP")) {
			codex.setTopp(config.get("dexter.topP", 1));
		}
		if (cfg.affectsConfiguration("dexter.frequencyPenalty")) {
			codex.setFreqPenalty(config.get("dexter.frequencyPenalty", 0));
		}
		if (cfg.affectsConfiguration("dexter.presencePenalty")) {
			codex.setPresPenalty(config.get("dexter.presencePenalty", 0));
		}
		if (cfg.affectsConfiguration("dexter.bestOf")) {
			codex.setBestOf(config.get("dexter.bestOf", 1));
		}
		if (cfg.affectsConfiguration("dexter.stopSequences")) {
			codex.setStopSequences(config.get("dexter.stopSequences", ""));
		}
	}));

	context.subscriptions.push(
		vscode.commands.registerCommand('dexter.addCodexContext', () => {
			const editor = vscode.window.activeTextEditor;
			if (!editor) {
				vscode.window.showInformationMessage("No open editor");
			} else {
				var selections = editor.selections;
				selections.forEach((selection, index) => {
					const text = editor.document.getText(selection);
					if (text.length > 0) {
						codex.addContext(text);
					} else {
						selections.splice(index, 1);
						return;
					}
				});
				// vscode.window.showInformationMessage(`Added ${selections.length} ${selections.length === 1 ? 'line' : 'lines'}. Total: ${codex.getContext().split('\n')}`)
				vscode.window.showInformationMessage(`Added context from ${selections.length} selections`)
			}
		})
	);

	context.subscriptions.push(
		vscode.commands.registerCommand('dexter.clearCodex', () => {
			codex.clearContext();
			codex.clearQueries();
			codex.clearStopSequences();
		})
	)

	function appendText(text: string) {
		vscode.commands.registerTextEditorCommand('dexter.insertText', function (editor, edit, args) {
			edit.insert(editor.selection.end, text)
		})
	}

	function replaceText(text: string) {
		vscode.commands.registerTextEditorCommand('dexter.insertText', function (editor, edit, args) {
			edit.replace(editor.selection.end, text)
		})
	} 

	context.subscriptions.push(
		vscode.commands.registerCommand('dexter.runCodeCompletion', async () => {
			const key = vscode.workspace.getConfiguration('dexter').get('apiKey', false);
			if (!key) {
				await vscode.window.showInputBox({
					title: "Set the OpenAI API key"
				}).then(input => {
					vscode.workspace.getConfiguration('dexter').update('apiKey', input);	
				})
			} else {
				const editor = vscode.window.activeTextEditor;
				if (!editor) {
					vscode.window.showInformationMessage("No open editor")
				} else {
					const selections = editor.selections;
					const replaceHighlight = vscode.workspace.getConfiguration('dexter').get('replaceHighlight');
					const sameCompletion = vscode.workspace.getConfiguration('dexter').get('sameCompletion');

					var insertText = replaceHighlight ? appendText : replaceText;
					//append if one or more selections contain characters
					var appendResult = !selections.every((value) => { return value.isEmpty; });
					
					await codex.complete().then(result => {
						if (result.includes('UNAUTHORIZED')){ //displayed if no output is generated (complete returns empty list)
							vscode.window.showErrorMessage('Bad API key');
						} else if (result.includes('ERROR')) {
							vscode.window.showInformationMessage('One or more errors occurred ')
						}
						selections.sort((a, b) => { 
							return a.active.compareTo(b.active);
						}).reverse().forEach((selection, index) => {
							codex.addQuery(editor.document.getText(selection));
							if (index === 0 || sameCompletion) {
								//result should ALWAYS be the same length as selections
								insertText(result[index]);
							} else {
								codex.complete().then(result => {
									insertText(result[index]);
								})
							}
						})
						return;
					});
				}
			}
		})
	);
//const line = editor.document.lineAt(selection.start.line).text;
	context.subscriptions.push(
		vscode.commands.registerCommand('dexter.setCodexResLength', async () => {
			await vscode.window.showInputBox({
				title: "Set the max number of tokens for codex to output (1-4096)",
				placeHolder: "64",
				validateInput: text => {
					const num = +text;
					if (num > 0 && num < 4097) {
						return text
					}
				}
			}).then(input => {
				if (input) {
					vscode.workspace.getConfiguration('dexter').update('responseLength', +input);
				} else {
					vscode.window.showErrorMessage("Response length must be a number"); 
				}
			});
		})
	);

	context.subscriptions.push(
		vscode.commands.registerCommand('dexter.setStopSequences', async () => {
			await vscode.window.showInputBox({
				title: "Up to 4 sequences (separated by /|. represent newline as ↵) that the model will end the output at.",
				placeHolder: ""
			}).then(input => {
				if (input) {
					vscode.workspace.getConfiguration('dexter').update('stopSequences', input);
				} else {
					vscode.workspace.getConfiguration('dexter').update('stopSequences', undefined);
				}
			});
		})
	);
}



// this method is called when your extension is deactivated
export function deactivate() {}
